In [1]:
#doing things with data
import pandas as pd
import itertools
import numpy as np
import scipy.stats as stats
from scipy.stats import pearsonr
import math

#imports for visualizations
import matplotlib.pyplot as plt
import seaborn as sns

#imports necessary for modeling
from sklearn.feature_selection import SelectKBest, f_regression
from sklearn.model_selection import GridSearchCV
from sklearn.feature_selection import RFE
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import RobustScaler
from sklearn.preprocessing import QuantileTransformer
from sklearn.linear_model import LassoLars
from sklearn.linear_model import TweedieRegressor
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import r2_score
from sklearn.metrics import r2_score, mean_squared_error
from sklearn.metrics import make_scorer

#import .py files
import wrangle as w
import explore as e
import modeling as m
import MPL_modeling_1 as mpl1
import MPL_modeling as mpl

#ignore warnings
import warnings
warnings.filterwarnings("ignore")



In [2]:
#import csv for exploration
df = w.get_data()
#split data into train, validate, test
train, test = e.split_data(df)

# Deep Learning Modeling with Multilayer Perceptron Model for Regression 
    

### The function below will run through a dictionary with the following MLP parameters creating and validating each model 
* Optimizers: rmsprop, adam
* Learning rates: .001, .01, .1
* Activations: relu, tanh
* Neurons: 8, 13, 21

* All models run through 5000 epochs

### The function will then predict using the best model and take the average of the RMSE and predicted revenue across 10 


#### Note: Due to  the stochastic nature of neural networks specific results will vary.

## Running the MLP on the Starbucks Data

In [3]:
# Get prediction data for Starbucks
starbucks_revenue_prediction = m.get_starbucks_Q1_2023_data_for_prediction()
# Run scale data function for Starbucks 
X_train_scaled, X_test_scaled, y_train, y_test, X_train, starbucks_revenue_prediction_scaled = m.starbucks_scaled_df(train, test, starbucks_revenue_prediction)
#calls the function to run the MLP model
mpl.auto_perceptron("Starbucks", X_train_scaled, y_train, X_test_scaled, y_test, starbucks_revenue_prediction_scaled)

Best MSE: 0.030
Best RMSE: 0.173
Best Parameters: {'optimizer': 'rmsprop', 'learning_rate': 0.01, 'activation': 'relu', 'neurons': 13}
1/1 [==============================] - 0s 85ms/step
This is the predictied next quarter revenue for Starbucks: [[9.143598]] billion dollars


In [4]:
#calls the function to run the MLP model 10 times and get the average RMSE and prediction for next quarter
mpl.predict_next_quarter_avg("Starbucks", X_train_scaled, X_test_scaled, y_test, y_train, starbucks_revenue_prediction_scaled)

1/1 [==============================] - 0s 55ms/step
This is the average predicted next quarter revenue for Starbucks: [[9.01923]] billion dollars
This is the average rmse for Starbucks: 0.28405689400172146 billion dollars


### Starbucks modeling summary
- Using a multilayer perceptron regression deep learning model on test data:
     - beat the ML model's baseline RMSE (2.08) with an average test RMSE of 0.29, 
- The final predicted revenue value for Starbucks 2023 Q2 is **9.0B**
     - fun fact, we checked if Starbucks had released their Q2 data.... Their reported revenue for 2023 Q2 will be released on the ***1st of Aug*** 

## Running the MLP on the Ford Data

In [5]:
# Get prediction data for Ford
ford_revenue_prediction = m.get_ford_Q1_2023_data_for_prediction()
# Scale data for modeling 
X_train_scaled, X_test_scaled, y_train, y_test, X_train, ford_revenue_prediction_scaled = m.ford_scaled_df(train, test,ford_revenue_prediction)
#calls the function to run the MLP model
mpl.auto_perceptron("Ford", X_train_scaled, y_train, X_test_scaled, y_test, ford_revenue_prediction_scaled)


Best MSE: 22.648
Best RMSE: 4.759
Best Parameters: {'optimizer': 'adam', 'learning_rate': 0.01, 'activation': 'relu', 'neurons': 8}
1/1 [==============================] - 0s 52ms/step
This is the predictied next quarter revenue for Ford: [[45.238247]] billion dollars


In [6]:
#calls the function to run the MLP model 10 times and get the average RMSE and prediction for next quarter
mpl.predict_next_quarter_avg("Ford", X_train_scaled, X_test_scaled, y_test, y_train, ford_revenue_prediction_scaled)

1/1 [==============================] - 0s 51ms/step
This is the average predicted next quarter revenue for Ford: [[43.705772]] billion dollars
This is the average rmse for Ford: 5.57420093330736 billion dollars


### Ford Modeling summary
- Using a multilayer perceptron regression deep learning model test data
     - beat the ML regression model's baseline RMSE (9.46) and best performing ML model's score (6.21) with an average test RMSE score of 5.57.
- The final predicted revenue value for Ford 2023 Q2 is **43.70B**
   - fun fact, we will verify this number on ***27 July 2023***, when Ford releases their Q2 earnings. 

## Running the MLP on the AT&T Data

In [7]:
# Get prediction data for ATT
att_revenue_prediction = m.get_att_Q1_2023_data_for_prediction()
# Scale data for modeling 
X_train_scaled, X_test_scaled, y_train, y_test, X_train, att_revenue_prediction_scaled = m.att_scaled_df(train, test,att_revenue_prediction)
#calls the function to run the MLP model
mpl.auto_perceptron("AT&T", X_train_scaled, y_train, X_test_scaled, y_test, att_revenue_prediction_scaled)

Best MSE: 20.379
Best RMSE: 4.514
Best Parameters: {'optimizer': 'rmsprop', 'learning_rate': 0.001, 'activation': 'tanh', 'neurons': 21}
1/1 [==============================] - 0s 52ms/step
This is the predictied next quarter revenue for AT&T: [[30.822668]] billion dollars


In [8]:
#calls the function to run the MLP model 10 times and get the average RMSE and prediction for next quarter
mpl.predict_next_quarter_avg("AT&T", X_train_scaled, X_test_scaled, y_test, y_train, att_revenue_prediction_scaled)

1/1 [==============================] - 0s 53ms/step
This is the average predicted next quarter revenue for AT&T: [[31.020859]] billion dollars
This is the average rmse for AT&T: 5.763621749070326 billion dollars


### AT&T modeling summary
- Using a multilayer perceptron regression deep learning model 
    - beat the baseline regression model's RMSE (10.91) and the best performing ML model (6.12) with an average RMSE score of 5.76
- The average predicted revenue value for ATT 2023 Q2 is **31.02B**
     - Their reported revenue for 2023 Q2 is due to be released on ***26 July!!!!!*** 
     